### PR TITLE
Support retrieving known Peripherals

### PR DESCRIPTION
**Pull request by Supereg**
# Support retrieving Peripherals

## :recycle: Current situation & Problem

SpeziBluetooth is currently optimized for device discovery and ad-hoc connection establishment. However, when dealing with device pairing, one typically scans for nearby devices once and saving the peripheral identifier and perform long running connection attempts (e.g., calling [`retrievePeripherals(withidentifiers:)`](https://developer.apple.com/documentation/corebluetooth/cbcentralmanager/retrieveperipherals(withidentifiers:)) and calling connect()). Connecting a peripheral does not time out and is therefore the most efficient way of establishing Bluetooth connecting with a set of known Bluetooth devices.
To support this use case, some modifications have been made to the underlying SpeziBluetooth infrastructure. New mechanisms were introduced to retrieve known peripherals (`BluetoothManager/retrievePeripheral(for:with:)` and `Bluetooth/retrieveDevice(for:as:)`). These device instances can stay allocated and SpeziBluetooth automatically frees resources once the framework user deallocates the peripheral instances. Several changes to the lifecycle handling of `BluetoothPeripheral`s and `BluetoothDevice`s have been made to support this new interfaces (e.g., carefully managing when objects are kept as strong references and when to reuse objects when, e.g., the same instance is getting discovered at the same time).

### SpeziDevices

SpeziDevices is an upcoming library to encapsulates a lot of standardized device interactions. This PR is driven by a lot of requirements of this library.

## :gear: Release Notes 

* Support retrieving known peripherals using `BluetoothManager/retrievePeripheral(for:with:)` and `Bluetooth/retrieveDevice(for:as:)`.
* The `BluetoothViews` target was removed and integrated into the `SpeziDevicesUI` of the SpezIDevices framework (see https://github.com/StanfordSpezi/SpeziDevices/pull/1, **Breaking**).
* The `BluetoothServices` target was renamed to `SpeziBluetoothServices` for more consistency (**Breaking**).
* Add `accessory` discovery criteria.
* New `nearby` and `lastActivity` state properties for peripherals and devices.
* The peripheral name is now preferred with the `name` property and the `localName` property can now be accessed individually on a peripheral (**Breaking**).
* Internally restructure discovery state into an `DiscoverySession` for better code overview.
* Configuration options (like minimumRSSI and advertisementStaleInterval) are now passed directly to the `scanNearbyDevices(...)` methods and modifiers allowing for more dynamic configuration.
* Adding `Bluetooth/stateSubscription` and `BluetoothManager/stateSubscription` to receive an AsyncStream of `BluetoothState` changes useful when required to observe CBCentral changes. 
* Adding `CharacteristicAccessor/subscription` and `DeviceStateAccessor/subscription` properties to receive an AsyncStream of changes to the characteristic value or device state.
* `onChange(initial:perform:)` methods now must not be set up in the initializer anymore and must not create strong self references. Instead setup onChange handlers in the `configure()` method and make sure to weakly capture `self`. Calling `onChange(initial:perform:)` method in the initializer results in a runtime crash with the error message providing a migration guide.
* New `powerOn()` and `powerOff()` methods to manually control `CBCentralManager` allocation.
* Fixed an issue where characteristic access continuations where accidentally leaked.
* Add new `ManufacturerIdentifier` model supporting parsing of manufacturer data in the advertisement.
* Added `ConnectedDevices` environment model to retrieve all connected devices from the SwiftUI environment.
* Add `Codable` conformance to all Characteristics.

This PR contains breaking changes, requiring a major version bump.

## :books: Documentation
Documentation was updated to reflect changed symbols (README currently points to 404 pages). Retrieving devices was documented with a small code examples.

## :white_check_mark: Testing
Existing tests verify that refactoring did not break existing infrastructure. Additional test case was added to test retrieving peripherals and ensuring the new reference semantics do not break and peripherals are properly and expectedly deallocated.

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: Sources/SpeziBluetooth/Bluetooth.swift [hunk 1]
```diff
@@ -417,30 +547,120 @@ public actor Bluetooth: Module, EnvironmentAccessible, BluetoothScanner, Bluetoo
...
+        let device = device.init()
+        
+        let didInjectAnything = device.inject(peripheral: peripheral, using: self)
+        if didInjectAnything {
+            initializedDevices[uuid] = device.weaklyReference
+        } else {
+            logger.warning(
+                """
+                \(Device.self) is an empty device implementation. \
+                The same peripheral might be instantiated via multiple \(Device.self) instances if not device property wrappers like
```
Comment (PSchmiedmayer) [inline Sources/SpeziBluetooth/Bluetooth.swift:619]: Might be a typo here?
```suggestion
                The same peripheral might be instantiated via multiple \(Device.self) instances if no device property wrappers like
```

File: Sources/SpeziBluetooth/CoreBluetooth/Model/DiscoverySession.swift
Comment (PSchmiedmayer) [inline Sources/SpeziBluetooth/CoreBluetooth/Model/DiscoverySession.swift:1]: Nice way to abstract this in this type 👍

File: Sources/SpeziBluetooth/Model/PropertySupport/CharacteristicAccessor.swift [hunk 2]
```diff
@@ -114,43 +116,79 @@ extension CharacteristicAccessor where Value: ByteDecodable {
...
-            }
-
-            // We save the instance in the global registrar if its available.
-            // It will be available if we are instantiated through the Bluetooth module.
-            // This indirection is required to support self referencing closures without encountering a strong reference cycle.
-            closures.insert(for: configuration.objectId, closure: closure)
-            return
+            preconditionFailure(
+                """
+                Register onChange(perform:) inside the initializer is not supported anymore. \
```
Comment (PSchmiedmayer) [inline Sources/SpeziBluetooth/Model/PropertySupport/CharacteristicAccessor.swift:178]: Nice debug help to transition to the new approach 👍

**Review by PSchmiedmayer — APPROVED** [review]
Nice job with all the work in this PR @Supereg; all the changes will be nice additions to SpeziBluetooth 2.0 🚀

I only had a few minor comments and ideas; feel free to merge the PR after the merge checks are all green and successful. Let me know if you run into any issues with the CI setup and feel free to ping me on Slack 👍

File: Sources/SpeziBluetooth/Bluetooth.swift [hunk 3]
```diff
@@ -261,40 +314,53 @@ public actor Bluetooth: Module, EnvironmentAccessible, BluetoothScanner, Bluetoo
...
         }
     }
 
+    /// Request to power up the Bluetooth Central.
+    ///
+    /// This method manually instantiates the underlying Central Manager and ensure that it stays allocated.
+    /// Balance this call with a call to ``powerOff()``.
+    ///
+    /// - Note : The underlying `CBCentralManager` is lazily allocated and deallocated once it isn't needed anymore.
+    ///     This is used to delay Bluetooth permission prompts to the latest possible moment avoiding to unexpectedly display power alerts.
```
[the remark this answers is not in the document]
Comment (Supereg) [reply Sources/SpeziBluetooth/Bluetooth.swift:345]: Actually both. The first time is a permission alert the next times CoreBluetooth displays a power alert (see [CBCentralManagerOptionShowPowerAlertKey](https://developer.apple.com/documentation/corebluetooth/cbcentralmanageroptionshowpoweralertkey)).

[the remark this answers is not in the document]
Comment (Supereg) [reply Sources/SpeziBluetooth/Bluetooth.swift]: I update the language to be more specific about permission and power alerts.

File: Package.swift [hunk 4]
```diff
@@ -20,18 +28,16 @@ let package = Package(
...
+        .library(name: "SpeziBluetoothServices", targets: ["SpeziBluetoothServices"]),
         .library(name: "SpeziBluetooth", targets: ["SpeziBluetooth"])
     ],
     dependencies: [
-        .package(url: "https://github.com/StanfordSpezi/SpeziFoundation", from: "1.0.4"),
-        .package(url: "https://github.com/StanfordSpezi/Spezi", from: "1.3.0"),
-        .package(url: "https://github.com/StanfordSpezi/SpeziNetworking", from: "2.0.1"),
-        .package(url: "https://github.com/StanfordSpezi/SpeziViews", from: "1.3.0"),
+        .package(url: "https://github.com/StanfordSpezi/SpeziFoundation", from: "1.1.0"),
+        .package(url: "https://github.com/StanfordSpezi/Spezi", branch: "feature/externally-managed-modules"),
```
Comment (Supereg) [inline Package.swift:36]: Waiting for https://github.com/StanfordSpezi/Spezi/pull/106

File: Sources/SpeziBluetooth/Bluetooth.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (PSchmiedmayer) [reply Sources/SpeziBluetooth/Bluetooth.swift]: Thank you!

File: Tests/UITests/UITests.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -740,8 +751,16 @@
 			isa = XCRemoteSwiftPackageReference;
 			repositoryURL = "https://github.com/StanfordBDHG/XCTestExtensions.git";
 			requirement = {
-				kind = upToNextMinorVersion;
-				minimumVersion = 0.4.10;
+				branch = "feature/xctassert-throws-async";
```
Comment (Supereg) [inline Tests/UITests/UITests.xcodeproj/project.pbxproj:754]: tag release

File: Package.swift [hunk 6]
```diff
@@ -20,18 +28,17 @@ let package = Package(
...
-        .package(url: "https://github.com/StanfordSpezi/SpeziNetworking", from: "2.0.1"),
-        .package(url: "https://github.com/StanfordSpezi/SpeziViews", from: "1.3.0"),
+        .package(url: "https://github.com/StanfordSpezi/SpeziFoundation", from: "1.1.0"),
+        .package(url: "https://github.com/StanfordSpezi/Spezi", from: "1.4.0"),
+        .package(url: "https://github.com/StanfordSpezi/SpeziNetworking", from: "2.1.0"),
         .package(url: "https://github.com/apple/swift-nio.git", from: "2.59.0"),
-        .package(url: "https://github.com/apple/swift-collections.git", from: "1.0.4")
-    ],
+        .package(url: "https://github.com/apple/swift-collections.git", from: "1.0.4"),
+        .package(url: "https://github.com/StanfordBDHG/XCTestExtensions.git", branch: "feature/xctassert-throws-async")
```
Comment (Supereg) [inline Package.swift:40]: tag release

**Pull request merged**